core/relayer/relayer/relayer.go

```diff
@@ -0,0 +1,159 @@
+package main
+
+import (
+	"fmt"
+	//"strings"
+	"path/filepath"
+	"time"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/xuperchain/xuperchain/core/common"
+	"github.com/xuperchain/xuperchain/core/kv/kvdb"
+	relay "github.com/xuperchain/xuperchain/core/relayer"
+	relayerpb "github.com/xuperchain/xuperchain/core/relayer/pb"
+)
+
+type Relayer struct {
+	deliverBlockCmd *relay.DeliverBlockCommand
+	queryBlockCmd   *relay.QueryBlockCommand
+	storage         *relay.Storage
+	queryMeta       *relayerpb.QueryMeta
+}
+
+func (relayer *Relayer) GetLastQueryBlockHeight() int64 {
+	return relayer.queryMeta.GetLastQueryBlockHeight()
+}
+
+func (relayer *Relayer) SaveBlockLoop() {
+	// 获取下一个需要获取的区块高度
+	currHeight := relayer.queryMeta.GetLastQueryBlockHeight() + 1
+	fmt.Println("[relayer] the beignning height:", currHeight)
+	// 持续从原链获取区块头(按照高度)
+	for {
+		block, err := relayer.queryBlockCmd.FetchBlockFromSrcChain(currHeight)
+		// 如果是not found, 可以等待
+		// 如果是原链挂掉了, 需要重启换节点
+		if common.NormalizedKVError(err) == common.ErrKVNotFound {
+			fmt.Println("[relayer] fetch block from src chain failed, try to fetch agagin", "target height:", currHeight, "err:", err)
+			time.Sleep(time.Duration(3) * time.Second)
+			continue
+		} else if err != nil {
+			fmt.Println("[relayer] fetch block from src chain failed, try to fetch agagin", "target height:", currHeight, "err:", err)
+			time.Sleep(time.Duration(100) * time.Millisecond)
+			currHeight--
+			continue
+		}
+		// 相关字段置空
+		block.Transactions = nil
+		block.MerkleTree = nil
+		block.FailedTxs = nil
+
+		blockBuf, pbErr := proto.Marshal(block)
+		// 序列化数据有问题, 重新获取区块
+		if pbErr != nil {
+			fmt.Println("[relayer] proto.Marshal block failed, try to get again")
+			continue
+		}
+		saveErr := relayer.storage.PutBlockHeader(block.GetBlockid(), blockBuf)
+		if saveErr != nil {
+			fmt.Println("[relayer] put block failed, err:", saveErr)
+			panic("put block into storage error:" + saveErr.Error())
+		}
+		saveErr = relayer.storage.PutHeightBlockid(currHeight, block.GetBlockid())
```

Comment:
What if this block's preHash is not the same with the last saved block(means the source chain just walked to a fork)?

Reply:
Corrected.